alf/networks/mdq_critic_networks.py

```diff
@@ -0,0 +1,515 @@
+            action_ind = self._action_qt.action_to_ind(actions[..., i])
+            if i == 0:
+                action_value_i = self._batched_index_select(
+                    action_values_i[i], -1, action_ind.long())
+                Q_values[i] = action_value_i
+                # KL-divergence
+                Q_values[i] = Q_values[i] - alpha * self._log_pi_uniform_prior
+            else:
+                selected_trans_action_value_i = self._batched_index_select(
+                    trans_action_values_i[i], -1, action_ind.long())
+                Q_values[i] = Q_values[i - 1] + selected_trans_action_value_i
+                # KL-divergence
+                Q_values[i] = Q_values[i] - alpha * self._log_pi_uniform_prior
+
+        # [action_dim, B, n, 1] -> [B, n, action_dim]
+        Q_values = Q_values.squeeze(3).permute(1, 2, 0)
+        return Q_values, state
+
+    def _net_forward_individual(self, inputs, alpha, i, state=()):
+        """Individiual forwarding for a specified action dims for value computation.
+        Args:
+            inputs (torch.Tensor): a tensor of the shape [B*action_dim, d]
+                or [B*action_dim, n, d]
+                with the data for each action dimension concanated along the
+                first dim for parallel computation
+            alpha: the temperature used for the advantage computation
+            i (int): the specified action dim to perform forwarding
+        Returns:
+            action_values_i (torch.Tensor): a tensor of the shape [B, n, action_bin]
+            state: empty
+        """
+        inputs, _ = self._pre_encoding_nets[i](inputs)
+        action_values_i, state = self._mid_shared_encoding_nets(inputs)
+        action_values_i, state = self._post_encoding_nets[i](action_values_i)
+        return action_values_i, state
+
+    def _net_forward_parallel(self, inputs, alpha, batch_size, state=()):
+        """Parallel forwarding across action dims for value computation.
+        Args:
+            inputs (torch.Tensor): a tensor of the shape [B*action_dim, d]
+                or [B*action_dim, n, d]
+                with the data for each action dimension concanated along the
+                first dim for parallel computation
+            alpha: the temperature used for the advantage computation
+            batch_size: the size of the original batch without stacking
+                all action dimensions
+        Returns:
+            action_values (torch.Tensor): a tensor of the shape
+                [action_dim, B, n, action_bin]
+            state: empty
+        """
+
+        inputs, _ = self._pre_encoding_parallel_net(inputs)
+        action_values_mid, state = self._mid_shared_encoding_nets(inputs)
+        out_size = self._mid_shared_encoding_nets.output_spec.shape[-1]
+        # note here use out_size not action_bins; out_size is hidden size
+        # [B*action_dim, n, d] ->  [action_dim, B, n, d]
+        action_values_mid = action_values_mid.view(
```

Comment:
Feel this should be `[B*action_dim, n, d]` ->  `[B, action_dim, n, d]`

Reply:
Updated to ```[action_dim*B, n, d] -> [action_dim, B, n, d]```.